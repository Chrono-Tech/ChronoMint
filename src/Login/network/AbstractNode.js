import EventEmitter from 'events'
import SockJS from 'sockjs-client'
import Stomp from 'webstomp-client'

const TIMEOUT_BASE = 1000

export default class AbstractNode extends EventEmitter {
  constructor ({ api, socket, trace }) {
    super()
    this._api = api
    this._trace = trace
    this._socket = socket
    this._ws = null
    this._client = null
    this._timeout = TIMEOUT_BASE
    this._missedActions = []
  }

  trace () {
    if (!this._trace) {
      return
    }
    // eslint-disable-next-line
    console.log(...arguments)
  }

  _handleConnectionSuccess = () => {
    this._timeout = TIMEOUT_BASE
    const actions = this._missedActions
    for (const action of actions) {
      action()
    }
    this._missedActions = []
  }

  _handleConnectionTimeout = () => {
    this._timeout *= 2
    this.connect()
  }

  _handleConnectionError = (e) => {
    this.trace(`Failed to connect. Retry after ${this._timeout / 1000} seconds`, e)
    setTimeout(this._handleConnectionTimeout, this._timeout)
  }

  connect () {
    if (!this._socket) {
      return
    }
    this._ws = new SockJS(this._socket.baseURL)
    this._client = Stomp.over(this._ws, {
      heartbeat: false,
      debug: false,
    })
    this._client.connect(
      this._socket.user,
      this._socket.password,
      this._handleConnectionSuccess,
      this._handleConnectionError
    )
  }

  executeOrSchedule (action) {
    if (this._socket) {
      action()
    } else {
      this._missedActions.push(action)
    }
  }

  /**
   * @abstract
   * @param txid
   */
  // eslint-disable-next-line
  getTransactionInfo (txid) {
    throw new Error('Not implemented')
  }

  /**
   * @abstract
   * @param address
   */
  // eslint-disable-next-line
  getAddressInfo (address) {
    throw new Error('Not implemented')
  }

  /**
   * @abstract
   * @param address
   */
  // eslint-disable-next-line
  getAddressUTXOS (address) {
    throw new Error('Not implemented')
  }

  /**
   * @abstract
   * @param rawtx
   */
  // eslint-disable-next-line
  send (rawtx) {
    throw new Error('Not implemented')
  }
}
