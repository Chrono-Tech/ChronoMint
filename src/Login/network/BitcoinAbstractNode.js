import EventEmitter from 'events'

export default class BitcoinAbstractNode extends EventEmitter {

  trace () {
    if (this._trace) {
      // eslint-disable-next-line
      console.log.apply(console, arguments)
    }
  }

  // eslint-disable-next-line
  async getTransactionInfo (txid) {
    throw new Error('Not implemented')
  }

  // eslint-disable-next-line
  async getAddressInfo (address) {
    throw new Error('Not implemented')
  }

  // eslint-disable-next-line
  async getAddressUTXOS (address) {
    throw new Error('Not implemented')
  }

  // eslint-disable-next-line
  async send (rawtx) {
    throw new Error('Not implemented')
  }
}
