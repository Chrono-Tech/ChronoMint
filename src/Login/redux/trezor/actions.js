import trezorProvider from '../../network/TrezorProvider'
import web3Provider from '../../network/Web3Provider'
import networkService, { NETWORK_SET_ACCOUNTS } from '../../redux/network/actions'

export const TREZOR_SET_U2F = 'trezor/SET_U2F'
export const TREZOR_SET_ETH_APP_OPENED = 'trezor/SET_ETH_APP_OPENED'
export const TREZOR_FETCHING = 'trezor/FETCHING'
export const TREZOR_FETCHED = 'trezor/FETCHED'

export const initTrezor = () => async (dispatch) => {
  const isInited = await trezorProvider.init()
  dispatch({ type: TREZOR_SET_U2F, isU2F: trezorProvider.isU2F() })
  dispatch({ type: TREZOR_SET_ETH_APP_OPENED, isETHAppOpened: trezorProvider.isETHAppOpened() })
  return isInited
}

export const startTrezorSync = () => async (dispatch) => {
  await dispatch(initTrezor())
  trezorProvider.on('connection', (isETHAppOpened) => dispatch({ type: TREZOR_SET_ETH_APP_OPENED, isETHAppOpened }))
  return trezorProvider.sync()
}

export const stopTrezorSync = (isReset = false) => (dispatch) => {
  trezorProvider.stop()
  if (!isReset) {
    return
  }
  // reset state if we do not intent to login
  dispatch({ type: NETWORK_SET_ACCOUNTS, accounts: [] })
  networkService.selectAccount(null)
  dispatch({ type: TREZOR_FETCHED, isFetched: false })
}

export const fetchAccount = () => async (dispatch) => {
  dispatch({ type: TREZOR_FETCHING })
  const accounts = await trezorProvider.fetchAccount()
  if (!accounts) {
    dispatch({ type: TREZOR_FETCHED, isFetched: false })
    return
  }
  dispatch({ type: NETWORK_SET_ACCOUNTS, accounts })
  networkService.selectAccount(accounts[ 0 ])
  dispatch({ type: TREZOR_FETCHED, isFetched: true })
  // we do not need to watching eth app on login
  dispatch(stopTrezorSync())
}

export const loginTrezor = () => {
  const providerURL = networkService.getProviderURL()
  trezorProvider.setupAndStart(providerURL)
  web3Provider.setWeb3(trezorProvider.getWeb3())
  web3Provider.setProvider(trezorProvider.getProvider())
}
