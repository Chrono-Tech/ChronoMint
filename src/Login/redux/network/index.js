import network from './reducer'

export {
  network,
}
