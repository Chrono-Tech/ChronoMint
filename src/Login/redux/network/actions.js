import AbstractContractDAO from 'dao/AbstractContractDAO'
import contractsManagerDAO from 'dao/ContractsManagerDAO'
import EventEmitter from 'events'
import Web3 from 'web3'
import metaMaskResolver from '../../network/metaMaskResolver'
import { getNetworkById, getScannerById, LOCAL_ID } from '../../network/settings'
import uportProvider, { UPortAddress } from '../../network/uportProvider'
import web3Provider, { Web3Provider } from '../../network/Web3Provider'
import { utils } from '../../settings'

const { web3Converter } = utils

export const DUCK_NETWORK = 'network'

export const NETWORK_LOADING = 'network/LOADING'
export const NETWORK_SET_ACCOUNTS = 'network/SET_ACCOUNTS'
export const NETWORK_SELECT_ACCOUNT = 'network/SELECT_ACCOUNT'
export const NETWORK_ADD_ERROR = 'network/ADD_ERROR'
export const NETWORK_CLEAR_ERRORS = 'network/CLEAR_ERRORS'
export const NETWORK_SET_TEST_RPC = 'network/SET_TEST_RPC'
export const NETWORK_SET_TEST_METAMASK = 'network/SET_TEST_METAMASK'
export const NETWORK_SET_NETWORK = 'network/SET_NETWORK'
export const NETWORK_SET_PROVIDER = 'network/SET_PROVIDER'

const ERROR_NO_ACCOUNTS = 'Couldn\'t get any accounts! Make sure your Ethereum client is configured correctly.'

export const loading = (isLoading = true) => (dispatch) => {
  dispatch({ type: NETWORK_LOADING, isLoading })
}

export const addError = (error) => (dispatch) => {
  dispatch({ type: NETWORK_ADD_ERROR, error })
}

export const clearErrors = () => (dispatch) => {
  dispatch({ type: NETWORK_CLEAR_ERRORS })
}

class NetworkService extends EventEmitter {
  connectStore (store) {
    this._store = store
    this._dispatch = store.dispatch
  }

  createNetworkSession = (account, provider, network) => {
    if (!this._account) {
      this._account = account
    }

    if (!account || !provider || !network) {
      throw new Error('Wrong session arguments')
    }
    const accounts = this._store.getState().get('network').accounts || []
    if (!accounts.includes(account)) {
      throw new Error('Account not registered')
    }

    web3Provider.resolve()

    AbstractContractDAO.setup(account)

    // sync with session state
    // this unlock login
    // dispatch(createSession(account))
    this.emit('createSession', {
      account, provider, network, dispatch: this._dispatch,
    })
  }

  async destroyNetworkSession (lastURL, isReset = true) {
    await AbstractContractDAO.stopWholeWatching()
    AbstractContractDAO.resetWholeFilterCache()
    if (isReset) {
      // for tests
      web3Provider.reset()
    }

    this.emit('destroySession', { lastURL, dispatch: this._dispatch })
  }

  async checkLocalSession (account, providerURL) {
    const isTestRPC = await this.checkTestRPC(providerURL)
    // testRPC must be exists
    if (!isTestRPC || !account) {
      return false
    }

    const web3 = new Web3()
    web3Provider.setWeb3(web3)
    web3Provider.setProvider(new web3.providers.HttpProvider(providerURL || (`//${location.hostname}:8545`)))
    const accounts = await web3Provider.getAccounts()

    // account must be valid
    if (!accounts.includes(account)) {
      return false
    }

    // contacts and network must be valid
    const isDeployed = await this.checkNetwork()
    if (!isDeployed) {
      return false
    }

    // all tests passed
    return true
  }

  async checkNetwork () {
    const dispatch = this._dispatch
    dispatch(loading())
    const isDeployed = await contractsManagerDAO.isDeployed()
    if (!isDeployed) {
      dispatch({
        type: NETWORK_ADD_ERROR,
        error: 'Network is unavailable.',
      })
    }
    return isDeployed
  }

  selectProvider = (selectedProviderId) => {
    const dispatch = this._dispatch
    dispatch({ type: NETWORK_SET_NETWORK, networkId: null })
    dispatch({ type: NETWORK_SET_PROVIDER, selectedProviderId })
  }

  selectNetwork = (selectedNetworkId) => {
    this._dispatch({ type: NETWORK_SET_NETWORK, selectedNetworkId })
  }

  async loginUport () {
    const dispatch = this._dispatch
    const provider = uportProvider.getUportProvider()
    dispatch(loading())
    dispatch(clearErrors())
    web3Provider.setWeb3(provider.getWeb3())
    web3Provider.setProvider(provider.getProvider())
    const encodedAddress: string = await provider.requestAddress()
    const { network, address }: UPortAddress = uportProvider.decodeMNIDaddress(encodedAddress)
    dispatch(this.selectNetwork(web3Converter.hexToDecimal(network)))
    dispatch({ type: NETWORK_SET_ACCOUNTS, accounts: [address] })
    this.selectAccount(address)
  }

  async loadAccounts () {
    const dispatch = this._dispatch
    dispatch(loading())
    dispatch({ type: NETWORK_SET_ACCOUNTS, accounts: [] })
    try {
      const accounts = await web3Provider.getAccounts()
      if (!accounts || accounts.length === 0) {
        throw new Error(ERROR_NO_ACCOUNTS)
      }
      dispatch({ type: NETWORK_SET_ACCOUNTS, accounts })
      if (accounts.length === 1) {
        this.selectAccount(accounts[ 0 ])
      }
      dispatch(loading(false))
      return accounts
    } catch (e) {
      dispatch(addError(e.message))
    }
  }

  async restoreLocalSession (account) {
    this.selectProvider(LOCAL_ID)
    this.selectNetwork(LOCAL_ID)
    await this.loadAccounts()
    this.selectAccount(account)
  }

  selectAccount = (selectedAccount) => {
    this._dispatch({ type: NETWORK_SELECT_ACCOUNT, selectedAccount })
  }

  getScanner = (params) => getScannerById(...params)

  getProviderSettings = () => {
    const state = this._store.getState()
    const { selectedNetworkId, selectedProviderId, isLocal } = state.get('network')
    const network = getNetworkById(selectedNetworkId, selectedProviderId, isLocal)
    const { protocol, host } = network

    return {
      network,
      url: protocol ? `${protocol}://${host}` : `//${host}`,
    }
  }

  getProviderURL = () => {
    const state = this._store.getState()
    const { selectedNetworkId, selectedProviderId, isLocal } = state.get('network')
    const { protocol, host } = getNetworkById(selectedNetworkId, selectedProviderId, isLocal)
    return protocol ? `${protocol}://${host}` : `//${host}`
  }

  checkMetaMask = () => {
    metaMaskResolver
      .on('resolve', (isMetaMask) => {
        try {
          if (isMetaMask) {
            this._dispatch({ type: NETWORK_SET_TEST_METAMASK })
          }
        } catch (e) {
          // eslint-disable-next-line
          console.error(e)
        }
      })
      .start()
  }

  async checkTestRPC (providerUrl) {
    // http only
    if (window.location.protocol === 'https:') {
      return false
    }

    const web3 = new Web3()
    web3.setProvider(new web3.providers.HttpProvider(providerUrl || (`//${location.hostname}:8545`)))
    const web3Provider = new Web3Provider(web3)

    const isDeployed = await contractsManagerDAO.isDeployed(web3Provider)
    if (!isDeployed) {
      return false
    }

    this._dispatch({ type: NETWORK_SET_TEST_RPC })
    return true
  }

  login (account) {
    this._account = account
    this.emit('login', { account: this._account, dispatch: this._dispatch })
  }
}

export default new NetworkService()
