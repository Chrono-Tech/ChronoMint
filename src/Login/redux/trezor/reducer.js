import * as a from './actions'

const initialState = {
  isHttps: window.location.protocol === 'https:',
  isU2F: false,
  isETHAppOpened: false,
  isFetching: false,
  isFetched: false,
}

export default (state = initialState, action) => {
  switch (action.type) {
    case a.TREZOR_SET_U2F:
      return {
        ...state,
        isU2F: action.isU2F,
      }
    case a.TREZOR_SET_ETH_APP_OPENED:
      return {
        ...state,
        isETHAppOpened: action.isETHAppOpened,
      }
    case a.TREZOR_FETCHING:
      return {
        ...state,
        isFetching: true,
      }
    case a.TREZOR_FETCHED:
      return {
        ...state,
        isFetched: action.isFetched,
        isFetching: false,
      }
    default:
      return state
  }
}
