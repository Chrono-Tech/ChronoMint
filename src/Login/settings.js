import MnemonicGenerateIcon from 'assets/img/mnemonic-key-color.svg'
import colors from 'styles/themes/variables'
import inverted from 'styles/themes/inversed'
import LocaleDropDown from 'layouts/partials/LocaleDropDown'
import SessionStorage from "utils/SessionStorage"
import web3Converter from 'utils/Web3Converter'
import ls from 'utils/LocalStorage'
import { SESSION_CREATE, SESSION_DESTROY } from 'redux/session/actions'
import theme from 'styles/themes/default'


export const components = {
  LocaleDropDown
}

export const assets = {
  MnemonicGenerateIcon,
}

export const styles = {
  colors,
  inverted,
  theme
}

export const utils = {
  SessionStorage,
  web3Converter,
  ls
}

export const constants = {
  SESSION_CREATE,
  SESSION_DESTROY
}
