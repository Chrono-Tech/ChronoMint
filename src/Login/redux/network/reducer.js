import { providerMap, getNetworksByProvider } from 'Login/network/settings'
import * as actions from './actions'

const initialState = {
  isLoading: false,
  isLocal: false,
  accounts: [],
  selectedAccount: null,
  errors: [],
  providers: [
    providerMap.chronoBank,
    providerMap.infura,
    providerMap.metamask,
    providerMap.uport,
    providerMap.local
  ],
  selectedProviderId: null,
  networks: [],
  selectedNetworkId: null
}

export default (state = initialState, action) => {
  switch (action.type) {
    case actions.NETWORK_LOADING:
      return {
        ...state,
        isLoading: action.isLoading
      }
    case actions.NETWORK_SET_TEST_RPC:
      return {
        ...state,
        isLocal: true,
        providers: state.providers.map(item => item.id === providerMap.local.id
          ? {
            ...item,
            disabled: false
          }
          : item
        )
      }
    case actions.NETWORK_SET_TEST_METAMASK:
      return {
        ...state,
        providers: state.providers.map(item => item.id === providerMap.metamask.id
          ? {
            ...item,
            disabled: false
          }
          : item
        )
      }
    case actions.NETWORK_SET_NETWORK:
      return {...state, selectedNetworkId: action.selectedNetworkId}
    case actions.NETWORK_SET_PROVIDER:
      return {
        ...state,
        selectedProviderId: action.selectedProviderId,
        networks: getNetworksByProvider(action.selectedProviderId, state.isLocal)
      }
    case actions.NETWORK_SET_ACCOUNTS:
      return {...state, accounts: action.accounts}
    case actions.NETWORK_SELECT_ACCOUNT:
      return {...state, selectedAccount: action.selectedAccount}
    case actions.NETWORK_CLEAR_ERRORS:
      return {...state, errors: []}
    case actions.NETWORK_ADD_ERROR:
      return {
        ...state,
        isLoading: false,
        errors: [...state.errors, action.error]
      }
    default:
      return state
  }
}
