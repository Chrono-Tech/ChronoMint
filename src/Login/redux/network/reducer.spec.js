import reducer from './reducer'
import * as actions from './actions'
import { accounts } from 'specsInit'
import { providerMap, infuraNetworkMap, infuraLocalNetwork } from 'Login/network/settings'

const selectedAccount = accounts[2]

describe('network reducer', () => {
  it('should return initial state', () => {
    expect(reducer(undefined, {}))
      .toEqual({
        isLoading: false,
        isLocal: false,
        accounts: [],
        selectedAccount: null,
        errors: [],
        providers: [
          providerMap.chronoBank,
          providerMap.infura,
          providerMap.metamask,
          providerMap.uport,
          providerMap.local
        ],
        selectedProviderId: null,
        networks: [],
        selectedNetworkId: null
      })
  })

  it('should handle NETWORK_SET_TEST_RPC', () => {
    const initialState = {
      isLocal: false,
      providers: [providerMap.local]
    }
    expect(reducer(initialState, {type: actions.NETWORK_SET_TEST_RPC}))
      .toEqual({
        isLocal: true,
        providers: [{
          id: 6,
          name: 'Local',
          disabled: false
        }]
      })
  })

  it('should handle NETWORK_SET_TEST_METAMASK', () => {
    const initialState = {
      providers: [providerMap.metamask]
    }
    expect(reducer(initialState, {type: actions.NETWORK_SET_TEST_METAMASK}))
      .toEqual({
        providers: [{
          id: 1,
          name: 'Metamask/Mist',
          disabled: false
        }]
      })
  })

  it('should handle NETWORK_SET_NETWORK', () => {
    expect(reducer({}, {type: actions.NETWORK_SET_NETWORK, selectedNetworkId: 2}))
      .toEqual({
        selectedNetworkId: 2
      })
  })

  it('should handle NETWORK_SET_PROVIDER without local', () => {
    const initialState = {
      isLocal: false,
      selectedProviderId: null,
      providers: [providerMap.metamask],
      networks: []
    }
    expect(reducer(initialState, {
      type: actions.NETWORK_SET_PROVIDER,
      selectedProviderId: providerMap.infura.id
    })).toEqual({
      isLocal: false,
      selectedProviderId: providerMap.infura.id,
      providers: [providerMap.metamask],
      networks: infuraNetworkMap
    })
  })

  it('should handle NETWORK_SET_PROVIDER with local', () => {
    const initialState = {
      isLocal: true,
      providers: [providerMap.metamask],
      networks: [],
      selectedProviderId: null
    }
    const state = reducer(initialState, {type: actions.NETWORK_SET_TEST_RPC})
    const expectedNetworks = infuraNetworkMap.concat(infuraLocalNetwork)
    expect(reducer(state, {
      type: actions.NETWORK_SET_PROVIDER,
      selectedProviderId: providerMap.infura.id
    })).toEqual({
      isLocal: true,
      networks: expectedNetworks,
      selectedProviderId: providerMap.infura.id,
      providers: [providerMap.metamask]
    })
  })

  it('should handle NETWORK_SET_ACCOUNTS', () => {
    expect(reducer({}, {type: actions.NETWORK_SET_ACCOUNTS, accounts}))
      .toEqual({
        accounts
      })
  })

  it('should handle NETWORK_SELECT_ACCOUNT', () => {
    expect(reducer({}, {type: actions.NETWORK_SELECT_ACCOUNT, selectedAccount}))
      .toEqual({
        selectedAccount
      })
  })

  it('should handle NETWORK_ADD_ERROR', () => {
    const initialState = {errors: ['bug', 'warning']}
    expect(reducer(initialState, {type: actions.NETWORK_ADD_ERROR, error: 'feature'}))
      .toEqual({
        errors: ['bug', 'warning', 'feature'],
        isLoading: false
      })
  })

  it('should handle NETWORK_CLEAR_ERRORS', () => {
    const initialState = {errors: ['bug', 'warning']}
    expect(reducer(initialState, {type: actions.NETWORK_CLEAR_ERRORS}))
      .toEqual({
        errors: []
      })
  })
})
