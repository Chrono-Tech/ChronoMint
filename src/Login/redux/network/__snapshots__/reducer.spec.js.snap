// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`network reducer should handle NETWORK_ADD_ERROR 1`] = `
  "errors": {
    "0": bug,
    "1": warning,
    "2": feature,
  },
  "isLoading": false,

`;

exports[`network reducer should handle NETWORK_CLEAR_ERRORS 1`] = `
  "errors": {
  },

`;

exports[`network reducer should handle NETWORK_SELECT_ACCOUNT 1`] = `
  "selectedAccount": 0x64371881e22e7cf48041c4da7efe0c9b7789aac7,

`;

exports[`network reducer should handle NETWORK_SET_ACCOUNTS 1`] = `
  "accounts": {
    "0": 0xb7237dd7dacd9da8e489b57d4631fc33342ff75f,
    "1": 0x236060666dbed392bb1f0b00b25e7c4b9cdcc4d5,
    "2": 0x64371881e22e7cf48041c4da7efe0c9b7789aac7,
    "3": 0x559b99a79cea7e1a495a244ee6ac816c1fb4a0be,
    "4": 0x670dc68af4125b8137d706bdec4abd8950d86c20,
    "5": 0x3ca0cbc8c9e2d629b13cfe7106535f532fb59c85,
    "6": 0x4d5d7edfb5e7057a0096bc653a7022c06360b3b3,
    "7": 0xd9ed6a1a2a635772583a87f29a089be5696c8ea9,
    "8": 0x5bd0ac539f551edcbff8045f0694729128e63d7f,
    "9": 0x7cc4056ab6c5d80c4288d59ef2263254b1fe6cb6,
  },

`;

exports[`network reducer should handle NETWORK_SET_NETWORK 1`] = `
  "selectedNetworkId": 2,

`;

exports[`network reducer should handle NETWORK_SET_PROVIDER with local 1`] = `
  "isLocal": true,
  "networks": {
    "0": {
      "bitcoin": bitcoin,
      "host": mainnet.infura.io/PVe9zSjxTKIP3eAuAHFA,
      "id": 1,
      "name": Mainnet (production),
      "nem": Mainnet,
      "protocol": https,
      "scanner": {
        "0": https://etherscan.io,
        "1": https://api.etherscan.io,
      },
    },
    "1": {
      "bitcoin": testnet,
      "host": rinkeby.infura.io/PVe9zSjxTKIP3eAuAHFA,
      "id": 4,
      "name": Rinkeby (test network),
      "nem": Testnet,
      "protocol": https,
      "scanner": https://rinkeby.etherscan.io,
    },
    "2": {
      "bitcoin": testnet,
      "host": kovan.infura.io/PVe9zSjxTKIP3eAuAHFA,
      "id": 42,
      "name": Kovan (test network),
      "nem": Testnet,
      "protocol": https,
      "scanner": https://kovan.etherscan.io,
    },
    "3": {
      "bitcoin": testnet,
      "host": localhost:3000/web3/,
      "id": 9999999999,
      "name": Localhost,
      "nem": Testnet,
      "protocol": https,
    },
  },
  "providers": {
    "0": {
      "disabled": true,
      "id": 1,
      "name": Metamask/Mist,
    },
  },
  "selectedProviderId": 2,

`;

exports[`network reducer should handle NETWORK_SET_PROVIDER without local 1`] = `
  "isLocal": false,
  "networks": {
    "0": {
      "bitcoin": bitcoin,
      "host": mainnet.infura.io/PVe9zSjxTKIP3eAuAHFA,
      "id": 1,
      "name": Mainnet (production),
      "nem": Mainnet,
      "protocol": https,
      "scanner": {
        "0": https://etherscan.io,
        "1": https://api.etherscan.io,
      },
    },
    "1": {
      "bitcoin": testnet,
      "host": rinkeby.infura.io/PVe9zSjxTKIP3eAuAHFA,
      "id": 4,
      "name": Rinkeby (test network),
      "nem": Testnet,
      "protocol": https,
      "scanner": https://rinkeby.etherscan.io,
    },
    "2": {
      "bitcoin": testnet,
      "host": kovan.infura.io/PVe9zSjxTKIP3eAuAHFA,
      "id": 42,
      "name": Kovan (test network),
      "nem": Testnet,
      "protocol": https,
      "scanner": https://kovan.etherscan.io,
    },
  },
  "providers": {
    "0": {
      "disabled": true,
      "id": 1,
      "name": Metamask/Mist,
    },
  },
  "selectedProviderId": 2,

`;

exports[`network reducer should handle NETWORK_SET_TEST_METAMASK 1`] = `
  "isMetamask": true,
  "providers": {
    "0": {
      "disabled": true,
      "id": 1,
      "name": Metamask/Mist,
    },
  },

`;

exports[`network reducer should handle NETWORK_SET_TEST_RPC 1`] = `
  "isLocal": true,
  "providers": {
    "0": {
      "disabled": true,
      "id": 6,
      "name": Local,
    },
  },

`;

exports[`network reducer should return initial state 1`] = `
  "accounts": {
  },
  "errors": {
  },
  "isLoading": false,
  "isLocal": false,
  "isMetamask": false,
  "networks": {
  },
  "preferMainnet": false,
  "priority": {
    "0": 4,
    "1": 2,
  },
  "providers": {
    "0": {
      "disabled": false,
      "id": 4,
      "name": ChronoBank,
    },
    "1": {
      "disabled": false,
      "id": 2,
      "name": Infura,
    },
  },
  "selectedAccount": null,
  "selectedNetworkId": null,
  "selectedProviderId": null,

`;
